src/main/java/com/oath/halodb/HaloDBInternal.java

```diff
@@ -281,6 +285,91 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
         }
     }
 
+    public boolean takeSnapShot() {
+        synchronized (snapshotLock) {
+            logger.info("Start generating the snapshot");
+
+            try {
+                try {
+                    compactionManager.pauseCompactionThread();
+                } catch (IOException e) {
+                    logger.warn("IOException when pause the compaction thread", e);
+                    return false;
+                }
+
+                while (compactionManager.isCompactionRunning()) {
+                    try {
+                        Thread.sleep(1000);
+                    } catch (InterruptedException e) {
+                        logger.info("The snapshot operation was interrupted");
+                        return false;
+                    }
+                }
+
+                File[] filesToCopy = dbDirectory.getPath().toFile()
+                    .listFiles(file -> {
+                        return (Constants.DATA_FILE_PATTERN.matcher(file.getName()).matches()
+                                || Constants.INDEX_FILE_PATTERN.matcher(file.getName()).matches()
+                                || Constants.TOMBSTONE_FILE_PATTERN.matcher(file.getName()).matches());
+                    });
+
+                logger.info("file number need to be copied: {}", filesToCopy.length);
+
+                try {
+                    Path dbDirectoryPath = dbDirectory.getPath();
+                    File snapshotDir = Paths.get(dbDirectoryPath.toFile().getAbsolutePath(), "snapshot").toFile();
+                    // Only support one snapshot now
+                    // TODO: support multiple snapshots if needed
+                    if (snapshotDir.exists()) {
+                        logger.warn("The snapshot dir is already existed. Delete the old one.");
+                        FileUtils.deleteDirectory(snapshotDir);
+                    }
+
+                    FileUtils.createDirectoryIfNotExists(snapshotDir);
+                    logger.info("Created directory for snapshot {}", snapshotDir.toString());
+
+                    for (File file : filesToCopy) {
+                        if (file.exists()) {
+                            Path dest = Paths.get(snapshotDir.getAbsolutePath(), file.getName());
+                            logger.debug("Create links from file {} to {}", file.getName(),
+                                         dest.toFile().getAbsoluteFile());
+                            Files.createLink(dest, file.toPath());
+                        } else {
+                            logger.warn("File {} maybe deleted by compaction job", file.getName());
+                        }
+                    }
+                } catch (IOException e) {
```

Comment:
Try to avoid using so many ```try/catch``` for code legibility.  I would suggest simplify to:
```java
try {
    // do the work
    writeLock.lock();
    try { 
     // do the work
    } finally {
        writeLock.unlock();
    }
    // do other work
} catch(IOException e) {
// log error and return false;
} finally {
    compactionManager.resumeCompaction();
}
```

Reply:
To be consistent with existing public methods, let exception thrown out instead of catching them might be another choice